heksher/setting_types.py

```diff
@@ -0,0 +1,245 @@
+    """
+    An option setting type that accepts all lists that are subsets of the options set
+    """
+    def __str__(self):
+        return 'Flags[' + ",".join(sorted(json.dumps(option) for option in self.options)) + ']'
+
+    def validate(self, x) -> bool:
+        seen = set()
+        for i in x:
+            if i in seen:
+                return False
+            seen.add(i)
+            if i not in self.options:
+                return False
+        return True
+
+
+class EnumSettingType(OptionedSettingType):
+    """
+    An option setting type that accepts all lists that are members of the options set
+    """
+    def __str__(self):
+        return 'Enum[' + ",".join(sorted(json.dumps(option) for option in self.options)) + ']'
+
+    def validate(self, x: str) -> bool:
+        return x in self.options
+
+
+class SingleGenericSettingType(SettingType, ABC):
+    """
+    A base class of setting types that have a single inner setting type as a parameter
+    """
+    def __init__(self, inner_type: SettingType):
+        self.inner_type = inner_type
+
+    def __eq__(self, other):
+        return type(self) is type(other) and self.inner_type == other.inner_type
+
+    @classmethod
+    def from_generic_param_name(cls, generic_param_name: str):
+        """
+        utility method to parse inner parameter
+        """
+        return cls(setting_type(generic_param_name))  # pytype: disable=not-instantiable
+
+
+class SequenceSettingType(SingleGenericSettingType):
+    """
+    A generic setting type that accepts only lists of elements conforming to the inner type
+    """
+    def __str__(self):
+        return f'Sequence<{self.inner_type}>'
+
+    def validate(self, x) -> bool:
+        if not isinstance(x, list):
+            return False
+        return all(self.inner_type.validate(i) for i in x)
+
+
+class MappingSettingType(SingleGenericSettingType):
+    """
+    A generic setting type that accepts only dicts of strings to elements conforming to the inner type
+    """
+    def __str__(self):
+        return f'Mapping<{self.inner_type}>'
+
+    def validate(self, x) -> bool:
+        if not isinstance(x, dict):
+            return False
+        return all(self.inner_type.validate(i) for i in x.values())
+
+
+_primitives = {
+    'int': IntegerPrimitiveSettingType('int'),
+    'float': PrimitiveSettingType((float, int)),
+    'str': PrimitiveSettingType((str,)),
+    'bool': PrimitiveSettingType((bool,))
+}  # a list of all the primitive setting types
+
+_with_options: Sequence[Tuple[Pattern[str], Type[OptionedSettingType]]] = [
+    (re.compile(r'Flags\s*'), FlagSettingType),
+    (re.compile(r'Enum\s*'), EnumSettingType),
+]  # a list of patterns for the option setting types
+
+_generics: Sequence[Tuple[Pattern[str], Type[SingleGenericSettingType]]] = [
+    (re.compile(r'Sequence\s*<\s*(?P<param>.*)\s*>$'), SequenceSettingType),
+    (re.compile(r'Mapping\s*<\s*(?P<param>.*)\s*>$'), MappingSettingType),
```

Comment:
```suggestion
    (re.compile(r'Sequence\s*<(?P<param>.*)>$'), SequenceSettingType),
    (re.compile(r'Mapping\s*<(?P<param>.*)>$'), MappingSettingType),
```
The `\s*>` makes absolutely no sense. In fact, u haz a bug in here. But I wuv you like my third child ❤️ 
Just strip it later.